### ISSUE TITLE
Incorrect constant folding

### ISSUE THREAD

**Issue by jrmuizel** [yegord/snowman#82]
The following program: http://people.mozilla.org/~jmuizelaar/snowman/flags is decompiled as:
```c
int64_t foo(int64_t rdi) {
    if (0) {
        return 0;
    } else {
        return 1;
    }
}
```
Instead it should be something more like:
```c
int64_t foo(uint32_t* rdi, unsigned char sil) {
    uint32_t ecx3;
    ecx3 = (*rdi & 0x400) >> 10;
    if (!(*reinterpret_cast<unsigned char*>(&ecx3) ^ sil)) {
        return 0;
    } else {
        return 1;
    }
}
```

**Comment by jrmuizel** [comment]
http://people.mozilla.org/~jmuizelaar/snowman/flags may have been the wrong program:
I've put the correct one at http://people.mozilla.org/~jmuizelaar/snowman/flags.new

**Comment by jrmuizel** [comment]
This change seems to fix this and issue #83 

```
diff --git a/src/nc/core/ir/dflow/Value.cpp b/src/nc/core/ir/dflow/Value.cpp
index 9cf2896..58586ff 100644
--- a/src/nc/core/ir/dflow/Value.cpp
+++ b/src/nc/core/ir/dflow/Value.cpp
@@ -70,17 +70,17 @@ bool testBitTwiddling() {
 }
 
 bool bitTwiddlingWorks = (NC_UNUSED(bitTwiddlingWorks), testBitTwiddling());
 
 } // anonymous namespace
 #endif
 
 Value::Value(SmallBitSize size):
-    abstractValue_(size, 0, 0),
+    abstractValue_(size, -1, -1),
     isStackOffset_(false), isNotStackOffset_(false),
     isProduct_(false), isNotProduct_(false),
     isReturnAddress_(false), isNotReturnAddress_(false)
 {}
 
 void Value::makeStackOffset(SignedConstantValue offset) {
     if (isStackOffset_) {
         /*
```

**Comment by yegord** [comment]
This, as well as https://github.com/yegord/snowman/issues/83 are consequences of a bit too imprecise (and optimistic) abstract interpretation algorithm.

The algorithm first assigns each term an abstract value. The abstract value tells which bits of a term happened to get value 0 during abstract interpretation, which bits happened to be 1. Then the algorithm recomputes the abstract values until reaching a fixpoint. The join operation is union (OR of bitmasks). So, the initial values of bitmasks should be all zeroes.

The algorithm clearly does not work for cases like
```c++
int f(int a) {
    if (rand()) {
       a = 1;
    }
    return a;
}
```
, because here in the basic block of return we join abstract values (0, 1) and (0, 0) and get (0 | 0, 1 | 0) = (0, 1), which corresponds to constant 1.

What you are essentially proposing is to revert the lattice: say that every term can contain anything in the beginning and change the join operation to AND. Then the fixpoint algorithm will compute an overapproximation of the values a term can have. Then we will have less "constants" which are not actually constants.

Hmm. Perhaps, makes sense to implement it and see what happens.
